Sources/SourceKitLSP/Swift/CodeActions/ConvertJSONToCodableStruct.swift

```diff
@@ -0,0 +1,287 @@
+    return
+      """
+      \(raw: self.buildStruct(from: jsonDictionary))
+      """
+  }
+
+  private static func buildStruct(
+    named name: String = "JSONValue",
+    at depth: Int = 0,
+    from jsonDictionary: Dictionary<String, Any>
+  ) -> String {
+    let members = self.buildStructMembers(at: depth + 1, from: jsonDictionary)
+    return
+      """
+      \(String(repeating: " ", count: depth * 2))struct \(name): Codable {
+      \(members.joined(separator: "\n"))
+      \(String(repeating: " ", count: depth * 2))}
+      """
+  }
+
+  private static func buildStructMembers(
+    at depth: Int,
+    from jsonDictionary: Dictionary<String, Any>
+  ) -> [String] {
+    var members = [String]()
+
+    var uniquer = Set<String>()
+    var nestedTypes = [(String, Dictionary<String, Any>)]()
+    func addNestedType(_ name: String, object: Dictionary<String, Any>) {
+      guard uniquer.insert(name).inserted else {
+        return
+      }
+      nestedTypes.append((name, object))
+    }
+
+    for key in jsonDictionary.keys.sorted() {
+      guard let value = jsonDictionary[key] else {
+        continue
+      }
+
+      let type = self.jsonType(of: value, suggestion: key)
+      if case let .object(name) = type, let subObject = value as? Dictionary<String, Any> {
+        addNestedType(name, object: subObject)
+      } else if case let .array(innerType) = type,
+        let array = value as? [Any],
+        let (nesting, elementType) = innerType.outermostObject
+      {
+        if let subObject = array.unwrap(nesting) {
+          addNestedType(elementType, object: subObject)
+        }
+      }
+
+      let member = "\(String(repeating: " ", count: depth * 2))var \(key): \(type.stringValue)"
+      members.append(member)
+    }
+
+    if !nestedTypes.isEmpty {
+      members.append("")
+      for (name, nestedType) in nestedTypes {
+        members.append(self.buildStruct(named: name, at: depth, from: nestedType))
+      }
+    }
+    return members
+  }
+
+  indirect enum JSONType {
+    case string
+    case double
+    case array(JSONType)
+    case null
+    case object(String)
+
+    var outermostObject: (Int, String)? {
+      var unwraps = 1
+      var value = self
+      while case let .array(inner) = value {
+        value = inner
+        unwraps += 1
+      }
+
+      if case let .object(typeName) = value {
+        return (unwraps, typeName)
+      } else {
+        return nil
+      }
+    }
+
+    var stringValue: String {
+      switch self {
+      case .string:
+        return "String"
+      case .double:
+        return "Double"
+      case .array(let ty):
+        return "[\(ty.stringValue)]"
+      case .null:
+        return "Void?"
+      case let .object(name):
+        return name
+      }
+    }
+  }
+
+  private static func jsonType(of value: Any, suggestion name: String) -> JSONType {
+    switch Swift.type(of: value) {
+    case is NSString.Type:
+      return .string
```

Comment:
Do we need to use `Swift.type(of:)` here? Can’t we just use

```suggestion
    switch value {
    case is NSString.self:
      return .string
```